ghcide/src/Development/IDE/Spans/AtPoint.hs

```diff
@@ -152,9 +154,9 @@ typeRowToLoc (row:.info) = do
 
 documentHighlight
   :: Monad m
-  => HieASTs a
-  -> RefMap a
-  -> Position
+  => HieASTs a -- ^ get AST
+  -> RefMap a -- ^ get map of identifiers to spans
+  -> Position -- ^ get position
```

Comment:
What do you mean by "get" here?

Reply:
That the argument gets applied to a function. Or function takes an argument.

---

Simple `get` helps mentally read docs as text, the function gets this this and this, returns that. Terminology of course can be disputed, aka function applies to argument or the other way around, & do Haskell function return a value or other terms can be used.

Of course to say is: "bind the ... argument to the ... parameter of a function"
But for simple dirt-cheap basic explanation - a simple explanation of how arguments are treated semantically, how combination produces a result - documenting it arguments works great.

Since programming is not about objects, but about processes - I try to turn object naming documentation (which type signatures "as is" are) into process describing.

---

The main thing to explain in the function is how to think about `HieASTs` plural here. For the purpose of the function it is not important are there many ASTs or we pretend we have 1 giant complex AST of everything. So just `AST`.

Next question is what `RefMap` is & how to think about it in the scope of the function. Function reminds & at the same time simplifies to the reader what obscure `RefMap` name is about. That it is important to see that is it simply a map from some identifiers to some spans.

In that, we connected the reader that some AST is taken & then identifiers of it get mapped to spans.

Then the position is taken. Probably here it needs to be additionally cleared-out that that position determines an identifier that gets highlighted.

The return type of the function is self-explanatory, `DocumentHighlight` - is a great self-explanatory name whose semantic meaning directly corresponds with the agenda of the function.

Reply:
I mean, if `get` is nothing but for readability, then these comments are saying `AST` for `HieASTs`, `map of identifiers to spans` for `RefMap`, and `position` for `Position`. These are mostly information-less. Why do we even want to give a doc if it does not say more than the type?

1. If you want to emphasize such point, you should emphasize that. Otherwise, just saying `AST` is redundant.
2. `RefMap` type is possibly worth some explanations, but your explanation applies to the type, not this argument. If you want to comment something on the type, it's better to write it on the type.

Reply:
Easy understandability & readability for docs are important. Otherwise, if it to transmit information by Shannon boundary (a mathematically maximally dense form of information) was optimal for agenda - we'd would not needed to learn protocols to correspond, languages & then we can just send people to read the source code or the binary directly.

---

In the doc for function `RefMap` purposefully gets oversimplified.
Real type is:
```haskell
type RefMap a = Map.Map Identifier [(Span, IdentifierDetails a)]
```

But doc purposefully says it is "a map identifiers to spans", not mentioning anything about `IdentifierDetails`. Docs are to streamline understanding & so because highlighting happens not based on `IdentifierDetails` but based on spans - docs forget to mention details unimportant to understanding & agenda. As a result - docs forget to mention unimportant details, but a person looking-up the meaning of a function hopefully would understand/recap it through reading the tiny doc.

Any statement is an emphasis in itself. Then generalized type names from a remote location, function doc is closer in information & scope to the purpose of the function putting emphasis, people tend to read the docs of a function they have interest first & think in the scope of a function. Additionally, if doc successfully conveys the process/meaning - the successful system of mental meaning/understanding is the emphasis how to read/understand the type names in the scope of the function.

Reply:
> This does not justify commenting types on a function.
It does not comment types on the functions. Those are docs to arguments of a function, even Haddock internally calls them comments for arguments: https://haskell-haddock.readthedocs.io/en/latest/markup.html#constructors-and-record-fields, https://haskell-haddock.readthedocs.io/en/latest/markup.html#function-arguments.

It comments what argument are for & about, arguments happen to have/be related to types closely & so the type semantics are present in the documentation for arguments, but bears the function argument scope.

---

> If you think that this function is hard to understand without any comments, then it is more helpful to put a comment directly relevant to this function

The meaning of the function documentation is to be readable/looked-up from the external scope. If to hide explanations inside the code - it defeats the purpose of documentation.

---

By the way, because no suggestions were given - but weak points of documentation were found - documentation get new edition.

Reply:
> It comments what argument are for & about

`map of identifiers to spans`, the part of your comment, is exactly the explanation of the type. If you want to explain what the argument is for/about, you should at the very least mention why this function needs such an identifiers-to-span map.

> If to hide explanations inside the code

What do you mean? I haven't mentioned anything to do with "hide explanations". I'm literally suggesting you to *expose* the explanation by putting it as a function doc comment.

Reply:
It is already exposed. It is a Haddock argument documentation. It is just a part of documentation that is not supported by HLS so far & showing those docs is what I am working on & what this PR is an initial stage of.

Reply:
Again, what do you mean by "to hide"? I don't follow the context. It feels like you randomly mentioned that hiddeness.

Reply:
>  then it is more helpful to put a comment directly relevant to this function

I've read it as supplying them as comments in the function.

Reply:
Oh I see, sorry for confusing words. I meant as a doc comment of this function.

Reply:
Ok. Deleted the docs. Done.